paddle/fluid/framework/ir/graph.cc

```diff
@@ -44,27 +48,56 @@ void CheckProgram(const ProgramDesc &program) {
-            op->Type());
+        if (!FLAGS_enforce_when_check_program) {
+          PADDLE_ENFORCE(
+              visit.find(_INT(OpRole::kOptimize)) == visit.end(),
+              "Cannot add backward operator %s after optimize operator.",
+              op->Type());
+        } else {
+          if (visit.find(_INT(OpRole::kOptimize)) != visit.end()) {
+            LOG(ERROR)
+                << "Cannot add backward operator %s after optimize operator."
+                << op->Type();
+          }
+        }
         break;
       case _INT(OpRole::kForward) | _INT(OpRole::kLoss):
-        PADDLE_ENFORCE(visit.find(_INT(OpRole::kBackward) |
-                                  _INT(OpRole::kLoss)) == visit.end(),
-                       "Cannot add backward|loss operator before "
-                       "forward|loss operator %s.",
-                       op->Type());
-        PADDLE_ENFORCE(
-            visit.find(_INT(OpRole::kOptimize)) == visit.end(),
-            "Cannot add forward|loss operator %s after optimize operator.",
-            op->Type());
+        if (!FLAGS_enforce_when_check_program) {
+          PADDLE_ENFORCE(visit.find(_INT(OpRole::kBackward) |
+                                    _INT(OpRole::kLoss)) == visit.end(),
+                         "Cannot add backward|loss operator before "
+                         "forward|loss operator %s.",
+                         op->Type());
+          PADDLE_ENFORCE(
+              visit.find(_INT(OpRole::kOptimize)) == visit.end(),
+              "Cannot add forward|loss operator %s after optimize operator.",
+              op->Type());
+        } else {
+          if (visit.find(_INT(OpRole::kBackward) | _INT(OpRole::kLoss)) !=
+              visit.end()) {
+            LOG(ERROR) << "Cannot add backward|loss operator before "
+                       << "forward|loss operator %s." << op->Type();
+          }
+
+          if (visit.find(_INT(OpRole::kOptimize)) != visit.end()) {
+            LOG(ERROR) << "Cannot add forward|loss operator %s after optimize "
+                          "operator."
+                       << op->Type();
+          }
+        }
         break;
       case _INT(OpRole::kOptimize):
       case _INT(OpRole::kOptimize) | _INT(OpRole::kLRSched):
-        PADDLE_ENFORCE(visit.find(_INT(OpRole::kBackward)) != visit.end(),
-                       "Optimize operators %s must follow backward operator.",
-                       op->Type());
+        if (!FLAGS_enforce_when_check_program) {
+          PADDLE_ENFORCE(visit.find(_INT(OpRole::kBackward)) != visit.end(),
+                         "Optimize operators %s must follow backward operator.",
+                         op->Type());
+        } else {
+          if (visit.find(_INT(OpRole::kBackward)) == visit.end()) {
+            LOG(ERROR) << "Optimize operators %s must follow backward operator."
+                       << op->Type();
+          }
+        }
```

Comment:
You should use the flag to skip [`CheckProgram`](https://github.com/PaddlePaddle/Paddle/pull/14829/files#diff-2e5c77597541abca34ee662f78464558R118) directly.